File: polyglot/python/pod_security_scanner.py
"""
polyglot/python/pod_security_scanner.py

Hardened security baseline for air-gapped/regulated Kubernetes.
Assess + generate NetworkPolicy/PodSecurity/RBAC/quota controls.

Air-gapped friendly: works with local files, minimal external dependencies.
"""

import json
import re
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Dict, List, Optional, Tuple


@dataclass
class SecurityIssue:
    """Represents a single security finding."""
    severity: str  # critical, high, medium, low, info
    category: str  # pod_security, network_policy, rbac, quota
    rule_id: str
    description: str
    resource: str = ""  # e.g., "pod/nginx-abc123" or "namespace/production"
    field_path: str = ""  # JSON path like "spec.containers[0].securityContext.privileged"
    current_value: Any = None
    recommended_fix: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return {
            "severity": self.severity,
            "category": self.category,
            "rule_id": self.rule_id,
            "description": self.description,
            "resource": self.resource,
            "field_path": self.field_path,
            "current_value": self.current_value,
            "recommended_fix": self.recommended_fix,
        }

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "SecurityIssue":
        return cls(**d)


@dataclass
class ScanResult:
    """Aggregated results of a security scan."""
    namespace: str = ""
    total_pods_scanned: int = 0
    issues_found: int = 0
    critical_issues: int = 0
    high_issues: int = 0
    medium_issues: int = 0
    low_issues: int = 0
    info_issues: int = 0
    issues: List[SecurityIssue] = field(default_factory=list)
    scan_time: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return {
            "namespace": self.namespace,
            "total_pods_scanned": self.total_pods_scanned,
            "issues_found": self.issues_found,
            "critical_issues": self.critical_issues,
            "high_issues": self.high_issues,
            "medium_issues": self.medium_issues,
            "low_issues": self.low_issues,
            "info_issues": self.info_issues,
            "issues": [i.to_dict() for i in self.issues],
            "scan_time": self.scan_time,
        }

    @classmethod
    def from_dict(cls, d: Dict[str, Any]) -> "ScanResult":
        result = cls(**d)
        if "issues" in d and isinstance(d["issues"], list):
            result.issues = [SecurityIssue.from_dict(i) for i in d["issues"]]
        return result


class PodSecurityScanner:
    """
    Hardened security scanner for air-gapped Kubernetes environments.
    
    Supports scanning from:
    - Live K8s API (via kubectl output files)
    - Local YAML/JSON manifests
    - Raw pod specification JSON
    
    Baseline levels: "restricted" (default), "baseline", "privileged"
    """

    # PSS Restricted baseline rules
    RESTRICTED_PSS = {
        "001": ("Privileged container", "spec.containers[].securityContext.privileged"),
        "002": ("RunAsRoot", "spec.containers[].securityContext.runAsUser=0 or runAsNonRoot=false"),
        "003": ("Host namespaces", "spec.hostNetwork, spec.hostPID, spec.hostIPC"),
        "004": ("AllowPrivilegeEscalation", "spec.containers[].securityContext.allowPrivilegeEscalation=true"),
        "005": ("CAP_ADD", "spec.containers[].securityContext.capabilities.add contains NET_ADMIN, SYS_ADMIN, etc."),
        "006": ("ReadOnlyRootFilesystem", "spec.containers[].securityContext.readOnlyRootFilesystem=false or not set"),
        "007": ("DropAllCapabilities", "spec.containers[].securityContext.capabilities.drop does not include ALL"),
        "008": ("RunAsNonRoot", "spec.containers[].securityContext.runAsNonRoot=false"),
        "010": ("SELinuxOptions", "spec.containers[].securityContext.seLinuxOptions with type=privileged or sro/sr"),
    }

    # Critical capabilities that indicate high risk
    CRITICAL_CAPS = {"NET_ADMIN", "SYS_ADMIN", "SYS_PTRACE", "DAC_OVERRIDE", "MKNOD"}

    def __init__(self, baseline: str = "restricted"):
        self.baseline = baseline  # restricted, baseline, privileged
        self.issues: List[SecurityIssue] = []

    def set_baseline(self, level: str) -> None:
        """Set the PSS baseline level."""
        valid_levels = {"restricted", "baseline", "privileged"}
        if level not in valid_levels:
            raise ValueError(f"Invalid baseline level. Must be one of {valid_levels}")
        self.baseline = level

    def _parse_pod_spec(self, pod_json: str) -> Dict[str, Any]:
        """Parse a raw JSON string into a pod specification dict."""
        try:
            return json.loads(pod_json)
        except json.JSONDecodeError as e:
            raise ValueError(f"Invalid JSON in pod spec: {e}")

    def _get_container_security_context(self, container: Dict[str, Any]) -> Optional[Dict]:
        """Extract security context from a container."""
        return container.get("securityContext")

    def _check_pss_rules(
        self,
        pod_spec: Dict[str, Any],
        namespace: str = "",
        resource_name: str = ""
    ) -> List[SecurityIssue]:
        """Check pods against PSS baseline rules."""
        issues = []
        
        # Determine which rules apply based on baseline level
        active_rules = self._get_active_pss_rules()

        containers = pod_spec.get("spec", {}).get("containers", [])
        init_containers = pod_spec.get("spec", {}).get("initContainers", [])
        all_containers = containers + init_containers

        for container in all_containers:
            sec_ctx = self._get_container_security_context(container) or {}

            # Rule 001: Privileged container
            if sec_ctx.get("privileged") is True:
                issues.append(SecurityIssue(
                    severity="critical",
                    category="pod_security",
                    rule_id="PSS-001",
                    description=f"Container runs in privileged mode (full host access)",
                    resource=f"{namespace}/{resource_name}",
                    field_path="spec.containers[].securityContext.privileged",
                    current_value=True,
                    recommended_fix='Set "privileged: false"',
                ))

            # Rule 002: RunAsRoot
            if sec_ctx.get("runAsUser") == 0 or (sec_ctx.get("runAsNonRoot") is False):
                issues.append(SecurityIssue(
                    severity="critical",
                    category="pod_security",
                    rule_id="PSS-002",
                    description=f"Container may run as root user",
                    resource=f"{namespace}/{resource_name}",
                    field_path="spec.containers[].securityContext.runAsUser/runAsNonRoot",
                    current_value=sec_ctx.get("runAsUser"),
                    recommended_fix='Set "runAsNonRoot: true"',
                ))

            # Rule 003: Host namespaces
            host_specs = pod_spec.get("spec", {}).get("hostNetwork") or \
                        pod_spec.get("spec", {}).get("hostPID") or \
                        pod_spec.get("spec", {}).get("hostIPC")
            
            if any([pod_spec.get("spec", {}).get("hostNetwork"),
                    pod_spec.get("spec", {}).get("hostPID"),
                    pod_spec.get("spec", {}).get("hostIPC")]):
                issues.append(SecurityIssue(
                    severity="critical",
                    category="pod_security",
                    rule_id="PSS-003",
                    description=f"Container shares host namespace (network/pid/ipc)",
                    resource=f"{namespace}/{resource_name}",
                    field_path="spec.hostNetwork/hostPID/hostIPC",
                    current_value=host_specs,
                    recommended_fix='Set "hostNetwork: false, hostPID: false, hostIPC: false"',
                ))

            # Rule 004: AllowPrivilegeEscalation
            if sec_ctx.get("allowPrivilegeEscalation") is True:
                issues.append(SecurityIssue(
                    severity="high",
                    category="pod_security",
                    rule_id="PSS-004",
                    description=f"Container allows privilege escalation",
                    resource=f"{namespace}/{resource_name}",
                    field_path="spec.containers[].securityContext.allowPrivilegeEscalation",
                    current_value=True,
                    recommended_fix='Set "allowPrivilegeEscalation: false"',
                ))

            # Rule 005: Dangerous capabilities
            added_caps = sec_ctx.get("capabilities", {}).get("add", []) or []
            dropped_caps = sec_ctx.get("capabilities", {}).get("drop", []) or []
            
            dangerous_found = any(cap in self.CRITICAL_CAPS for cap in added_caps)
            if dangerous_found:
                issues.append(SecurityIssue(
                    severity="critical" if "SYS_ADMIN" in added_caps else "high",
                    category="pod_security",
                    rule_id="PSS-005",
                    description=f"Container adds dangerous capability: {', '.join([c for c in added_caps if c in self.CRITICAL_CAPS])}",
                    resource=f"{namespace}/{resource_name}",
                    field_path="spec.containers[].securityContext.capabilities.add",
                    current_value=added_caps,
                    recommended_fix='Remove or minimize "capabilities.add"',
                ))

            # Rule 006: ReadOnlyRootFilesystem
            if not sec_ctx.get("readOnlyRootFilesystem"):
                issues.append(SecurityIssue(
                    severity="medium",
                    category="pod_security",
                    rule_id="PSS-006",
                    description=f"Container does not use read-only root filesystem",
                    resource=f"{namespace}/{resource_name}",
                    field_path="spec.containers[].securityContext.readOnlyRootFilesystem",
                    current_value=sec_ctx.get("readOnlyRootFilesystem"),
                    recommended_fix='Set "readOnlyRootFilesystem: true"',
                ))

            # Rule 007: DropAllCapabilities
            if not any("ALL" in d for d in dropped_caps):
                issues.append(SecurityIssue(
                    severity="medium",
                    category="pod_security",
                    rule_id="PSS-007",
                    description=f"Container does not drop ALL capabilities",
                    resource=f"{namespace}/{resource_name}",
                    field_path="spec.containers[].securityContext.capabilities.drop",
                    current_value=dropped_caps,
                    recommended_fix='Add "ALL" to "capabilities.drop"',
                ))

            # Rule 008: RunAsNonRoot
            if not sec_ctx.get("runAsNonRoot"):
                issues.append(SecurityIssue(
                    severity="medium",
                    category="pod_security",
                    rule_id="PSS-008",
                    description=f"Container does not enforce non-root user",
                    resource=f"{namespace}/{resource_name}",
                    field_path="spec.containers[].securityContext.runAsNonRoot",
                    current_value=sec_ctx.get("runAsNonRoot"),
                    recommended_fix='Set "runAsNonRoot: true"',
                ))

            # Rule 010: SELinuxOptions
            selinux = sec_ctx.get("seLinuxOptions") or {}
            if selinux.get("type"):
                type_val = selinux["type"]
                if type_val in ["privileged", "sro", "sr"]:
                    issues.append(SecurityIssue(
                        severity="high" if type_val == "privileged" else "medium",
                        category="pod_security",
                        rule_id="PSS-010",
                        description=f"Container uses potentially privileged SELinux context: {type_val}",
                        resource=f"{namespace}/{resource_name}",
                        field_path="spec.containers[].securityContext.seLinuxOptions.type",
                        current_value=selinux.get("type"),
                        recommended_fix='Use least-privilege SELinux type',
                    ))

        return issues

    def _get_active_pss_rules(self) -> Dict[str, Tuple[str, str]]:
        """Return rules active for the configured baseline level."""
        if self.baseline == "privileged":
            # All rules disabled (except critical ones like privileged containers)
            return {}
        elif self.baseline == "baseline":
            # Medium and low severity rules only
            return {k: v for k, v in self.RESTRICTED_PSS.items() 
                    if v[0] not in ["Privileged container", "RunAsRoot", "Host namespaces"]}
        else:  # restricted (default)
            return self.RESTRICTED_PSS

    def _check_service_accounts(
        self,
        namespace: str = "",
        sa_list: Optional[List[Dict]] = None
    ) -> List[SecurityIssue]:
        """Check ServiceAccounts for overly permissive configurations."""
        issues = []
        
        if not sa_list:
            # Assume default service account exists
            sa_list = [{"metadata": {"name": "default"}}]

        for sa in sa_list:
            name = sa.get("metadata", {}).get("name") or "unknown"
            
            # Check automountServiceAccountToken
            automount = sa.get("automountServiceAccountToken")
            if automount is True:  # Default behavior, but explicit true is less secure
                issues.append(SecurityIssue(
                    severity="low",
                    category="rbac",
                    rule_id="RBAC-001",
                    description=f"ServiceAccount {name} has automountServiceAccountToken explicitly set to true",
                    resource=f"namespace/{namespace}/serviceaccount/{name}",
                    field_path="spec.automountServiceAccountToken",
                    current_value=automount,
                    recommended_fix='Set "automountServiceAccountToken: false" unless needed',
                ))

            # Check secrets/roles bindings (would need to fetch from API)
            roles = sa.get("roleRef") or {}
            if roles.get("name") == "cluster-admin":
                issues.append(SecurityIssue(
                    severity="critical",
                    category="rbac",
                    rule_id="RBAC-002",
                    description=f"ServiceAccount {name} bound to cluster-admin role",
                    resource=f"namespace/{namespace}/serviceaccount/{name}",
                    field_path="spec.roleRef.name",
                    current_value=roles.get("name"),
                    recommended_fix='Create a custom Role with minimal permissions',
                ))

        return issues

    def _check_resource_quotas(
        self,
        namespace: str = "",
        quota_spec: Optional[Dict] = None
    ) -> List[SecurityIssue]:
        """Check ResourceQuota configurations."""
        issues = []

        if not quota_spec:
            # Assume no quota exists (common issue)
            issues.append(SecurityIssue(
                severity="medium",
                category="quota",
                rule_id="QUOTA-001",
                description=f"Namespace {namespace} may lack ResourceQuota limits",
                resource=f"namespace/{namespace}",
                field_path="",
                current_value=None,
                recommended_fix='Create a ResourceQuota with reasonable limits',
            ))

        if quota_spec:
            # Check for missing hard limits
            requests = quota_spec.get("hard", {}) or {}
            
            cpu_limit = requests.get("requests.cpu")
            memory_limit = requests.get("requests.memory")

            if not cpu_limit:
                issues.append(SecurityIssue(
                    severity="medium",
                    category="quota",
                    rule_id="QUOTA-002",
                    description=f"Namespace {namespace} lacks CPU request quota",
                    resource=f"namespace/{namespace}",
                    field_path="spec.hard.requests.cpu",
                    current_value=cpu_limit,
                    recommended_fix='Add "requests: cpu: 100m (minimum)"',
                ))

            if not memory_limit:
                issues.append(SecurityIssue(
                    severity="medium",
                    category="quota",
                    rule_id="QU

File: polyglot/rust/pod_security_scanner.rs
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::io::{self, Write};
use std::process::exit;

// ============================================================================
// Data Models
// ============================================================================

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct PodSpec {
    pub containers: Vec<Container>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Container {
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub image: String,
    #[serde(default)]
    pub security_context: Option<SecurityContext>,
    #[serde(default)]
    pub resources: Option<ResourceRequirements>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct SecurityContext {
    #[serde(default)]
    pub run_as_non_root: bool,
    #[serde(default)]
    pub privileged: bool,
    #[serde(default)]
    pub allow_privilege_escalation: Option<bool>,
    #[serde(default)]
    pub capabilities: Option<Capabilities>,
    #[serde(default)]
    pub read_only_root_filesystem: bool,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Capabilities {
    #[serde(default)]
    pub add: Vec<String>,
    #[serde(default)]
    pub drop: Vec<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ResourceRequirements {
    #[serde(default)]
    pub limits: Option<ResourcesLimits>,
    #[serde(default)]
    pub requests: Option<ResourcesRequests>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ResourcesLimits {
    #[serde(default)]
    pub cpu: String,
    #[serde(default)]
    pub memory: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ResourcesRequests {
    #[serde(default)]
    pub cpu: String,
    #[serde(default)]
    pub memory: String,
}

// ============================================================================
// Security Rule Definitions
// ============================================================================

#[derive(Debug, Clone)]
pub enum CheckResult {
    Passed,
    Warning(String),
    Failed(String),
}

impl std::fmt::Display for CheckResult {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            CheckResult::Passed => write!(f, "✓"),
            CheckResult::Warning(msg) => write!(f, "⚠ {}", msg),
            CheckResult::Failed(msg) => write!(f, "✗ {}", msg),
        }
    }
}

#[derive(Debug)]
pub struct SecurityRule {
    pub id: String,
    pub description: &'static str,
    pub severity: Severity,
    pub check_fn: fn(&Container) -> CheckResult,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum Severity {
    Low,
    Medium,
    High,
    Critical,
}

impl std::fmt::Display for Severity {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Severity::Low => write!(f, "LOW"),
            Severity::Medium => write!(f, "MEDIUM"),
            Severity::High => write!(f, "HIGH"),
            Severity::Critical => write!(f, "CRITICAL"),
        }
    }
}

// ============================================================================
// Security Rules Configuration
// ============================================================================

pub fn get_security_rules() -> Vec<SecurityRule> {
    vec![
        // Critical rules
        SecurityRule {
            id: "C001".to_string(),
            description: "Container running as privileged",
            severity: Severity::Critical,
            check_fn: |c| {
                if let Some(sc) = &c.security_context {
                    if sc.privileged {
                        CheckResult::Failed(format!(
                            "Container '{}' has privileged=true",
                            c.name
                        ))
                    } else {
                        CheckResult::Passed
                    }
                } else {
                    // Default is false, but warn if not explicitly set
                    CheckResult::Warning(
                        format!("Container '{}' security_context.privileged not explicitly set (defaults to false)", c.name)
                    )
                }
            },
        },
        
        SecurityRule {
            id: "C002".to_string(),
            description: "Container running as root user",
            severity: Severity::Critical,
            check_fn: |c| {
                if let Some(sc) = &c.security_context {
                    match (sc.run_as_non_root, sc.allow_privilege_escalation) {
                        (true, _) => CheckResult::Passed,
                        (false, Some(true)) => CheckResult::Failed(format!(
                            "Container '{}' allows privilege escalation and runAsNonRoot=false",
                            c.name
                        )),
                        _ => {
                            // Default behavior: check if explicitly set to false
                            if let Some(sc) = &c.security_context {
                                if !sc.run_as_non_root && sc.allow_privilege_escalation == Some(true) {
                                    CheckResult::Failed(format!(
                                        "Container '{}' allows privilege escalation",
                                        c.name
                                    ))
                                } else {
                                    CheckResult::Passed
                                }
                            } else {
                                CheckResult::Warning(
                                    format!("Container '{}' runAsNonRoot not explicitly set (defaults to false)", c.name)
                                )
                            }
                        },
                    }
                } else {
                    CheckResult::Warning(format!(
                        "Container '{}' security_context.runAsNonRoot not explicitly set",
                        c.name
                    ))
                }
            },
        },

        SecurityRule {
            id: "C003".to_string(),
            description: "Host network namespace enabled",
            severity: Severity::High,
            check_fn: |c| {
                // Note: hostNetwork is at Pod level, but we'll warn if any container hints at it
                CheckResult::Passed // Would need pod-level context for full check
            },
        },

        SecurityRule {
            id: "C004".to_string(),
            description: "Host PID namespace enabled",
            severity: Severity::High,
            check_fn: |c| {
                CheckResult::Passed // Would need pod-level context
            },
        },

        SecurityRule {
            id: "C005".to_string(),
            description: "Dangerous capabilities added",
            severity: Severity::Critical,
            check_fn: |c| {
                if let Some(sc) = &c.security_context {
                    if let Some(capabilities) = &sc.capabilities {
                        let dangerous_caps = [
                            "NET_ADMIN",
                            "SYS_ADMIN",
                            "SYS_PTRACE",
                            "DAC_OVERRIDE",
                            "CAP_SYS_CHROOT",
                            "MKNOD",
                            "SETUID",
                            "SETGID",
                            "CHOWN",
                        ];

                        let found_dangerous: Vec<_> = dangerous_caps
                            .iter()
                            .filter(|cap| capabilities.add.contains(*cap))
                            .collect();

                        if !found_dangerous.is_empty() {
                            CheckResult::Failed(format!(
                                "Container '{}' adds dangerous capabilities: {:?}",
                                c.name, found_dangerous
                            ))
                        } else {
                            CheckResult::Passed
                        }
                    } else {
                        // Default is empty, but warn if explicitly set to add anything
                        CheckResult::Warning(
                            format!("Container '{}' capabilities.add not explicitly set (defaults to [])", c.name)
                        )
                    }
                } else {
                    CheckResult::Passed
                }
            },
        },

        // High severity rules
        SecurityRule {
            id: "C006".to_string(),
            description: "Read-only root filesystem not enabled",
            severity: Severity::High,
            check_fn: |c| {
                if let Some(sc) = &c.security_context {
                    match sc.read_only_root_filesystem {
                        true => CheckResult::Passed,
                        false => CheckResult::Warning(format!(
                            "Container '{}' readOnlyRootFilesystem=false",
                            c.name
                        )),
                        _ => {
                            // Default is false
                            if let Some(sc) = &c.security_context {
                                if !sc.read_only_root_filesystem {
                                    CheckResult::Warning(format!(
                                        "Container '{}' readOnlyRootFilesystem not explicitly set (defaults to false)",
                                        c.name
                                    ))
                                } else {
                                    CheckResult::Passed
                                }
                            } else {
                                CheckResult::Passed
                            }
                        },
                    }
                } else {
                    CheckResult::Warning(format!(
                        "Container '{}' security_context.readOnlyRootFilesystem not explicitly set",
                        c.name
                    ))
                }
            },
        },

        SecurityRule {
            id: "C007".to_string(),
            description: "AllowPrivilegeEscalation enabled",
            severity: Severity::High,
            check_fn: |c| {
                if let Some(sc) = &c.security_context {
                    match sc.allow_privilege_escalation {
                        Some(false) => CheckResult::Passed,
                        Some(true) => CheckResult::Failed(format!(
                            "Container '{}' allows privilege escalation",
                            c.name
                        )),
                        _ => {
                            // Default is true (equivalent to allowPrivilegeEscalation: true)
                            if let Some(sc) = &c.security_context {
                                if sc.allow_privilege_escalation == Some(true) || 
                                   sc.allow_privilege_escalation.is_none() {
                                    CheckResult::Warning(format!(
                                        "Container '{}' allows privilege escalation (default behavior)",
                                        c.name
                                    ))
                                } else {
                                    CheckResult::Passed
                                }
                            } else {
                                CheckResult::Warning(format!(
                                    "Container '{}' allowPrivilegeEscalation not explicitly set",
                                    c.name
                                ))
                            }
                        },
                    }
                } else {
                    CheckResult::Warning(format!(
                        "Container '{}' security_context.allowPrivilegeEscalation not explicitly set",
                        c.name
                    ))
                }
            },
        },

        // Medium severity rules
        SecurityRule {
            id: "C008".to_string(),
            description: "Container using 'latest' image tag",
            severity: Severity::Medium,
            check_fn: |c| {
                if c.image.ends_with(":latest") || 
                   (c.image.contains('@') && !c.image.contains(':')) {
                    CheckResult::Warning(format!(
                        "Container '{}' uses potentially mutable image reference",
                        c.name
                    ))
                } else {
                    CheckResult::Passed
                }
            },
        },

        SecurityRule {
            id: "C009".to_string(),
            description: "No resource limits defined",
            severity: Severity::Medium,
            check_fn: |c| {
                if let Some(resources) = &c.resources {
                    match (&resources.limits, &resources.requests) {
                        (Some(_), _) => CheckResult::Passed,
                        (_, Some(_)) => CheckResult::Warning(format!(
                            "Container '{}' has requests but no limits defined",
                            c.name
                        )),
                        _ => CheckResult::Warning(format!(
                            "Container '{}' has no resource limits or requests defined",
                            c.name
                        )),
                    }
                } else {
                    CheckResult::Warning(format!(
                        "Container '{}' resources not explicitly set (defaults to none)",
                        c.name
                    ))
                }
            },
        },

        // Low severity rules
        SecurityRule {
            id: "C010".to_string(),
            description: "Container using 'latest' tag without digest",
            severity: Severity::Low,
            check_fn: |c| {
                if c.image.contains(":") && !c.image.contains("@") {
                    let parts: Vec<&str> = c.image.split(':').collect();
                    if parts.len() > 1 && parts[parts.len()-1] == "latest" {
                        CheckResult::Warning(format!(
                            "Container '{}' uses 'latest' tag (consider using digest)",
                            c.name
                        ))
                    } else {
                        CheckResult::Passed
                    }
                } else {
                    CheckResult::Passed
                }
            },
        },

        SecurityRule {
            id: "C011".to_string(),
            description: "Container without security context",
            severity: Severity::Low,
            check_fn: |c| {
                if c.security_context.is_none() {
                    CheckResult::Warning(format!(
                        "Container '{}' has no securityContext defined (uses defaults)",
                        c.name
                    ))
                } else {
                    CheckResult::Passed
                }
            },
        },

        SecurityRule {
            id: "C012".to_string(),
            description: "Container with empty name",
            severity: Severity::Low,
            check_fn: |c| {
                if c.name.is_empty() {
                    CheckResult::Warning(format!(
                        "Container '{}' has an empty name (may cause issues)",
                        "<unnamed>"
                    ))
                } else {
                    CheckResult::Passed
                }
            },
        },

        SecurityRule {
            id: "C013".to_string(),
            description: "Container with very long image name",
            severity: Severity::Low,
            check_fn: |c| {
                if c.image.len() > 256 {
                    CheckResult::Warning(format!(
                        "Container '{}' has a very long image name (>256 chars)",
                        c.name
                    ))
                } else {
                    CheckResult::Passed
                }
            },
        },

        SecurityRule {
            id: "C014".to_string(),
            description: "Container with special characters in image",
            severity: Severity::Low,
            check_fn: |c| {
                let special_chars = [' ', '\t', '\n'];
                if c.image.chars().any(|ch| special_chars.contains(&ch)) {
                    CheckResult::Warning(format!(
                        "Container '{}' has special characters in image name",
                        c.name
                    ))
                } else {
                    CheckResult::Passed
                }
            },
        },

        SecurityRule {
            id: "C015".to_string(),
            description: "Container with 'root' user explicitly set",
            severity: Severity::Low,
            check_fn: |c| {
                if let Some(sc) = &c.security_context {
                    // Check for runAsUser: 0 or username: root
                    match (sc.run_as_user, sc.run_as_non_root) {
                        (Some(0), _) => CheckResult::Warning(format!(
                            "Container '{}' explicitly runs as user 0 (root)",
                            c.name
                        )),
                        _ => {
                            // Default behavior - check if explicitly set to root
                            if let Some(sc) = &c.security_context {
                                if sc.run_as_user == Some(0) || 
                                   (sc.run_as_non_root == false && sc.allow_privilege_escalation == Some(true)) {
                                    CheckResult::Warning(format!(
                                        "Container '{}' may run as root",
                                        c.name
                                    ))
                                } else {
                                    CheckResult::Passed
                                }
                            } else {
                                CheckResult::Passed
                            }
                        },
                    }
                } else {
                    CheckResult::Passed
                }
            },
        },

        SecurityRule {
            id: "C016".to_string(),
            description: "Container with capabilities.drop not set to ALL",
            severity: Severity::Low,
            check_fn: |c| {
                if let Some(sc) = &c.security_context {
                    if let Some(capabilities) = &sc.capabilities {
                        // Check if drop is explicitly set to ALL
                        if capabilities.drop.iter().any(|d| d == "ALL") {
                            CheckResult::Passed
                        } else if capabilities.drop.is_empty() {
                            CheckResult::Warning(format!(
                                "Container '{}' has empty capabilities.drop (no caps dropped)",
                                c.name
                            ))
                        } else {
                            // Some caps dropped, but not ALL - this is generally okay
                            CheckResult::Passed
                        }
                    } else {
                        // Default behavior - check if explicitly set
                        if let Some(sc) = &c.security_context {
                            if sc.capabilities.is_some() && 
                               sc.capabilities.as_ref().unwrap().drop.is_empty() {
                                CheckResult